bindings/python/src/tokenizer.rs

```diff
@@ -310,10 +326,27 @@ impl From<PyArrayUnicode> for tk::InputSequence<'_> {
 struct PyArrayStr(Vec<String>);
 impl FromPyObject<'_> for PyArrayStr {
     fn extract(ob: &PyAny) -> PyResult<Self> {
-        let array = ob.downcast::<PyArray1<u8>>()?;
-        let arr = array.as_array_ptr();
+        if unsafe { npyffi::PyArray_Check(ob.py(), ob.as_ptr()) } == 0 {
+            return Err(exceptions::PyTypeError::new_err("Expected an np.array"));
+        }
+        let arr = ob.as_ptr() as *mut npyffi::PyArrayObject;
+
+        if unsafe { (*arr).nd } != 1 {
+            return Err(exceptions::PyTypeError::new_err(
+                "Expected a 1 dimensional np.array",
+            ));
+        }
+        if unsafe { (*arr).flags }
+            & (npyffi::NPY_ARRAY_C_CONTIGUOUS | npyffi::NPY_ARRAY_F_CONTIGUOUS)
+            == 0
+        {
+            return Err(exceptions::PyTypeError::new_err(
+                "Expected a contiguous np.array",
+            ));
+        }
+        let n_elem = unsafe { *(*arr).dimensions } as usize;
+
         let (type_num, data) = unsafe { ((*(*arr).descr).type_num, (*arr).data) };
-        let n_elem = array.shape()[0];
 
         if type_num != 17 {
```

Comment:
I just noticed that this case is not about a Unicode array, but actually an array containing `PyObject` which we do support since 0.16, so this whole method be able to become safe by using `downcast::<PyArray1<PyObject>>()` and then `array.readonly().as_array().iter()` which would even remove the requirement of a contiguous array.